### PR TITLE
Combine step/accept for ModeReducer using MutableOption

### PR DESCRIPTION
**Pull request by charleso**

### CODE REVIEW

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducer.scala [hunk 1]
```diff
@@ -11,8 +12,8 @@ trait ModeReducer {
   /** My kingdom for decent rank-2 types */
   type X
   def seed: X
-  def step(state: X, fact: Fact): X
-  def accept(state: X, fact: Fact): Boolean
+  /** Return the next state if the step is valid, or none otherwise */
+  def step(state: X, option: MutableOption[X], fact: Fact): MutableOption[X]
```
Comment (charleso) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducer.scala]: Should this be renamed to `stepOption` or something?

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducerSpec.scala [hunk 2]
```diff
@@ -43,14 +44,8 @@ Construction
   def laws(mr: ModeReducer): Prop = new Properties("ModeReducer laws") {
     property("seed is always equal") =
         mr.seed ?= mr.seed
-    property("accept is consistent") = forAll((f: Fact) =>
-      mr.accept(mr.seed, f) ?= mr.accept(mr.seed, f)
-    )
     property("step is consistent") = forAll((f: Fact) =>
-      mr.step(mr.seed, f) ?= mr.step(mr.seed, f)
-    )
-    property("step and accept is consistent") = forAll((f: Fact) =>
```
Comment (charleso) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducerSpec.scala]: Is there a new law I can introduce now?

File: ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducer.scala (hunk 1 above)
Comment (markhibberd) [inline ivory-operation/src/main/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducer.scala]: I wouldn't. Think it is fine.

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducerSpec.scala (hunk 2 above)
Comment (markhibberd) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/extraction/mode/ModeReducerSpec.scala]: Not that I can see from this, kind of supersedes.

**Comment by markhibberd** [comment]
:palm_tree:

**Pull request merged**